src/backend/storage/lmgr/proc.c

```diff
@@ -1166,6 +1167,17 @@ ProcSleep(LOCALLOCK *locallock, LockMethod lockMethodTable)
 					RememberSimpleDeadLock(MyProc, lockmode, lock, proc);
 					early_deadlock = true;
 					break;
+				} else if (myProc_is_remotexact && proc->isRemoteXact) {
+					/* 
+					 * Remotexact 
+					 * The MyProc is executing a remotexact and waiting for
+					 * another process that is also executing a remotexact. So
+					 * set the early_deadlock flag to true and break out of the
+					 * loop. Also record the deadlock info for later message. 
+					 */
+					RememberSimpleDeadLock(MyProc, lockmode, lock, proc);
+					early_deadlock = true;
+					break;
```

Comment:
Does this mean we abort `MyProc` if `proc` is waiting for `MyProc` and both of them are remotexact? 

This might cause more false positives because if `proc` figured that `MyProc` is remotexact, `proc` would be already aborted before this. Otherwise, `MyProc` would eventually be blocked behind `proc` or some other remote transaction later and abort itself then. But if `MyProc` is never blocked behind another remote transaction then we don't have to abort anyone.

Reply:
This code is actually symmetric to lines 1157 - 1170. 
Consider T1(MyProc) and T2(proc). 
T1 (still a local xact) locked some data Y in the current_region
T2 was a remotexact, it locked some data X in the current_region. 
T2 now tries to lock Y and thus adds itself to the queue and waits for T1 to commit or abort. 
T2 wakes up and completes it DeadLockCheck we have not identified any deadlock because T1 is still local. 
Now T1 upgrades itself to a remotexact and then tries to lock X. At this point, there is a deadlock that we can detect early and simply abort T1, allowing T2 to proceed.

This code is basically an optimization to avoid that initial sleep before the deadlock check. Similar to the case of looking for single region deadlocks.

Reply:
I'm still not sure how this helped. Your example does not involve a distributed deadlock (all edges of the cycle are in the same region) and will be resolved by the 1157 - 1170 block.

Reply:
Sorry, that was not a good example. 

T1 is a local xact. 
T1 locks tuple A. 
T2 is a remote xact - (say a validation request, to make this clearer). 
T2 attempts to lock tuple A, and hence adds itself to the lock queue of T1.
T2 wakes up and carries out a Deadlock check - it passes because T1 is still local, and there are no other local cycles. 
T1 upgrades to a remotexact. 
T1 attempts to lock some tuple (B) - It does not really matter. Finds that it is already locked (say by T3, another local xact) and carries out this check before adding itself to the queue. 
At this point, we iterate through all processes waiting on T1. We find that T2 is remote and is waiting on T1, and therefore, we need to abort one of the two. Since we are past the deadlock check for  T2, it will have to be T1.

Reply:
I couldn't really determine a way to tell if the Deadlock check has been called. But in reality, it does not matter. We just make sure that we abort one of the two transactions involved in a potential multi-region deadlock.

Reply:
I think it is a bit too conservative if we do it this way. If T1 never gets into a deadlock with T2, we'll make an unnecessary abort. But if it does, we're still safe because T1 or some other multi-region transaction down the chain will eventually be aborted. 

We only benefit from early abort if a deadlock indeed happens but when it does not, which is the common case, we lose on too many aborts.

Reply:
I think this is a safety requirement. Let's discuss this in person today?